vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -417,49 +479,110 @@ and
                                                 optionsAssociation.Remove(projectContext) |> ignore
                                                 project.Disconnect()))
-                for referencedSite in ProjectSitesAndFiles.GetReferencedProjectSites (site, this.SystemServiceProvider) do
-                    let referencedProjectId = setup referencedSite                    
-                    project.AddProjectReference(ProjectReference referencedProjectId)
+            | _ -> ()
+            for referencedSite in ProjectSitesAndFiles.GetReferencedProjectSites (site, this.SystemServiceProvider) do
+                let referencedProjectId = setup referencedSite
+                project.AddProjectReference(ProjectReference referencedProjectId)
 
             projectId
         setup (siteProvider.GetProjectSite()) |> ignore
 
+    member private __.ProvideProjectSiteProvider(workspace:Workspace, project:Project) =
+        let visualStudioWorkspace = workspace :?> VisualStudioWorkspace
+        let hier = visualStudioWorkspace.GetHierarchy(project.Id)
+        {new IProvideProjectSite with
+             member this.GetProjectSite() =
+                 let compileItems () = projectInfoManager.GetSourceFiles(project.FilePath)
+                 let compilerFlags () = [| "" |]
+                 let caption () = project.Name
+                 let projFileName () = project.FilePath
+                 let taskProvider = None
+                 let taskReporter = None
+                 let targetFrameworkMoniker = ""
+                 let projectGuid () = project.Id.Id.ToString()
+                 let creationTime = System.DateTime.Now
+                 let assemblyReferences () = projectInfoManager.GetReferenceFiles(project.FilePath)
+                 { new Microsoft.VisualStudio.FSharp.LanguageService.IProjectSite with
+                     member __.SourceFilesOnDisk() = compileItems ()
+                     member __.DescriptionOfProject() = caption ()
+                     member __.CompilerFlags() = compilerFlags ()
+                     member __.ProjectFileName() = projFileName ()
+                     member __.ErrorListTaskProvider() = taskProvider
+                     member __.ErrorListTaskReporter() = taskReporter
+                     member __.AdviseProjectSiteChanges(_,_) = ()
+                     member __.AdviseProjectSiteCleaned(_,_) = ()
+                     member __.AdviseProjectSiteClosed(_,_) = ()
+                     member __.IsIncompleteTypeCheckEnvironment = false
+                     member __.TargetFrameworkMoniker = targetFrameworkMoniker
+                     member __.ProjectGuid = projectGuid ()
+                     member __.LoadTime = creationTime
+                     member __.ProjectProvider = Some this
+                     member __.AssemblyReferences() = assemblyReferences ()
+                 }
+         interface IVsHierarchy with
```

Comment:
Why is an ``IVsHierarchy`` implementation is needed here?  I do vaguely remember - is it that crappy code in ProjectSitesAndFiles.fs that searches for IVsHierarchy for project references?  However if that's the case, then let's document it, and let's also eventually get rid of this in favour of adding an appropriate method into IProjectSite, or even just using the existing ``AssemblyReferences`` method